### PR TITLE
Deprecate Tick.{gridOn,tick1On,label1On,...} in favor of set_visible.

### PR DESCRIPTION
**Pull request by anntzer**
Visibility attributes should just be set on the underlying artists
(`tick.gridline.set_visible`, etc.) rather than maintaining yet another
layer of control.

---

Note that this technically breaks code that overrode the `gridOn` etc. attributes using properties (such as the old implementation of skewed axes) but I can't think of a reasonable way to emit a warning for these...

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by efiring** [comment]
I have mixed feelings about this, so far.  The 'tick1On' etc. attributes are ugly but fairly deeply embedded.  Is this the time to make such an invasive change?  Or would it be just as well to defer it for a more thorough overhaul of the overly complex and sluggish Tick/Axis/Spine system?  Maybe we will never get to that, though.

**Comment by anntzer** [comment]
I don't think the system is *that* deeply embedded (but it has been around for a while, for sure).  The deprecation is basically just replacing the attributes by calls to set_visible throughout; the only complication is with code that proxies these attributes using properties (i.e., the skew axes example), but this kind of code is always going to interact in a tricky manner with deprecations (because we arguably can't cover all possible ways in which our attributes have been proxied...).

**Comment by tacaswell** [comment]
I do not disagree it simplifies things, but I am not sure it is worth the disruption in this case.

attn @astrofrog Does this cause issues for WCSAxes?

**Comment by anntzer** [comment]
Should work, but haven't tried :)

**Comment by astrofrog** [comment]
I don't think this will cause issues for WCSAxes, though it will require some updates in APLpy (https://github.com/aplpy/aplpy/blob/eb09ec39506db2da91f92a383016d1328df731c3/aplpy/colorbar.py#L133) but I can live with that.

**Comment by anntzer** [comment]
Not going to insist on marking this as RC, but still bumping the suggestion up.

**Comment by tacaswell** [comment]
I am ok with this in principle, needs a rebase and version updates.

**Comment by anntzer** [comment]
done

**Comment by timhoffm** [comment]
There are a number of cases (most notably `update_position()` and `draw()`) that were previously only executed conditionally on the visibility.

I agree that updating always is the more correct thing to do (otherwise the value is not correct when enabling visibility later on.). On the other hand, I'm a bit concerned of the performance impact. Have you evaluated this?

**Comment by anntzer** [comment]
I haven't, but I also think that it is the correct thing to do.  Obviously it would be possible to hide them back again under a check for get_visible() but the semantics would be so-slightly different (if someone was really relying on that) and I don't think it's a good feature to keep anyways...
Can probably do a proper benchmark at some point, if you feel it's necessary.

File: examples/specialty_plots/skewt.py [hunk 1]
```diff
@@ -24,66 +26,24 @@
...
-
-    @property
-    def label2On(self):
-        return self._label2On and self._need_upper()
-
-    @label2On.setter
-    def label2On(self, value):
-        self._label2On = value
+    def draw(self, renderer):
+        with ExitStack() as stack:
```
Comment (jklymak) [inline examples/specialty_plots/skewt.py]: This is kind of mysterious for an example w/o a comment why you need it.

Comment (anntzer) [reply examples/specialty_plots/skewt.py]: added an explanation, I hope it helps

**Comment by jklymak** [comment]
CI failures unrelated...

@efiring, you expressed misgivings.  Are you placated, or ??  If so, I'll merge...

**Comment by efiring** [comment]
OK to merge once it passes the tests.

**Comment by jklymak** [comment]
@anntzer after a rebase feel free to merge once it passes tests....

**Comment by anntzer** [comment]
Done (modulo appveyor which is queued, but there's no reason for it to fail :p).

**Comment by ImportanceOfBeingErnest** [comment]
> The ``label`` attribute, which was an alias for ``label1``, has been deprecated.

This will hit every call to `seaborn.heatmap`, which is a pretty common function inside seaborn.

**Comment by jklymak** [comment]
will it be a problem for seaborn to replace by `label1`?

**Comment by ImportanceOfBeingErnest** [comment]
I mean it used to be a simple alias, `self.label = self.label1`. So that shouldn't be a problem. I just wanted to make sure we are aware of the implication that once 3.1 is released thousands of seaborn users will get some warning message.

**Comment by timhoffm** [comment]
Good point. We shouldn't do that.

**Pull request merged**